worker

```diff
@@ -11,62 +52,96 @@ BH_UUID=${BH_UUID}
+# shellcheck disable=SC2164
+cd 1>&2 "$WORKDIR"
+# We have an EXIT trap which makes this code reachable:
+# shellcheck disable=SC2317
+function cleanup() {
+  echo "# cleaning ${WORKDIR}"
+  chmod --recursive u+w "$WORKDIR"
+  rm --recursive "$WORKDIR"
+} 1>&2
+trap cleanup EXIT
 
-cat <<"EOF"
-<!DOCTYPE html>
-<html lang=en>
-<title>a title</title>
-<p>a paragraph</p>
-<pre>
-EOF
-
-(
-WORKDIR="$(mktemp -d --tmpdir bids-hook."$BH_UUID".XXXXXXXXXXX)"
-cd "$WORKDIR"
-
-#git clone "$GITEA_REPO" dataset && cd dataset # this is standard but clones a full copy of the repo which might be expensive
-#git clone -b "${BH_COMMIT}" --depth 1 "$GITEA_REPO" dataset && cd dataset # this doesn't work unless BH_COMMIT is a branch name
-# this is more complicated but it allows to us to --depth 1 with branches OR commit IDs:
-(mkdir dataset && cd dataset && git init && git remote add origin "$GITEA_REPO" && git fetch --depth 1 origin "${BH_COMMIT}")
-#trap 'echo -n "cleaning "; pwd; chmod -R +w "$WORKDIR" && rm -rf "$WORKDIR"' EXIT
-(
-cd dataset
+# Check out the repository, redirecting all output to stderr.
+(exec 1>&2
+# We'd like to do 'git clone -b "$BH_COMMIT" --depth 1 "$GITEA_REPO"'
+# but maybe "$BH_COMMIT" is a commit ID instead of a branch name, so
+# we do it this way instead.
+echo "# cloning ${GITEA_REPO}"
+git init
+git remote add origin "$GITEA_REPO"
+git fetch --depth 1 origin "$BH_COMMIT"
 git checkout "$BH_COMMIT"
-set -e
-
-if git ls-remote --exit-code "$GITEA_REPO" "refs/heads/git-annex" >/dev/null; then
-  ANNEXED=1
-else
-  ANNEXED=""
-fi
 
-if [ -n "$ANNEXED" ]; then
-  set -x
+# If this is a git-annex repository, we need to get the contents.
+if git ls-remote --exit-code origin refs/heads/git-annex >/dev/null; then
+  echo '# getting git-annexed files'
   # this reduces copies; always overrides annex.hardlink even if that is set system-wide
   git config annex.thin true
   # make sure we don't corrupt origin accidentally
   git config remote.origin.annex-readonly true
   git config annex.private true # XXX this doesn't do anything until git-annex 10
   git annex init
   git annex dead here # this is like annex.private, but has to be run
-  git annex sync --only-annex --no-content # grab the git-annex branch (since we did a shallow clone above)
-  git annex copy --from origin  # NB: using copy --from origin and not git annex get to ensure we're validating the contents of origin and not any special remotes
-else
-  set -x
+  # grab the git-annex branch (since we did a shallow clone above)
+  git annex sync --only-annex --no-content
+  # NB: using 'copy --from origin' and not 'git annex get; to ensure we're
+  # validating the contents of origin and not any special remotes
+  git annex copy --from origin
 fi
-
-bids-validator --verbose .
 )
-) 2>&1
-R=$?
 
-cat <<"EOF"
+echo 1>&2 '# running bids-validator'
+OUTPUT=$(bids-validator .) && STATUS=$? || STATUS=$?
+WARNINGS=$(bids-validator . --json | jq '.issues.warnings | length') || true
```

Comment:
It's disappointing to have to run this twice.  `bids-validator` is pretty quick, even on large datasets, but I don't trust it to stay that way.

Is there another flag where maybe we could get the warnings routed...elsewhere?

Or can we patch bids-validator so that it exits:

- 0 if no issues exist
- 1 if issues exist and contain errors
- 2 if issues exist

Does it already do that?

Reply:
> Does it already do that?

I don't think so. Looking at the source, here is the meaning of each exit status for bids-validator:
* [Exit 3](https://github.com/bids-standard/bids-validator/blob/47d2912e91d1c78ac93470d5e76a0752ecfbce64/bids-validator/cli.js#L42-L50) for internal errors (the equivalent of a traceback).
* [Exit 2](https://github.com/bids-standard/bids-validator/blob/47d2912e91d1c78ac93470d5e76a0752ecfbce64/bids-validator/cli.js#L56-L62) for problems opening the dataset directory.
* [Exit 1](https://github.com/bids-standard/bids-validator/blob/47d2912e91d1c78ac93470d5e76a0752ecfbce64/bids-validator/cli.js#L66-L71) if the `.bids-validator-config.json` is invalid or the dataset contains "error"-level issues.
* [Exit 0](https://github.com/bids-standard/bids-validator/blob/47d2912e91d1c78ac93470d5e76a0752ecfbce64/bids-validator/cli.js#L73) otherwise, including if there are "warning"-level issues.